### ISSUE TITLE
toCcc' rule not firing with GHC 8.2.2

### ISSUE THREAD

**Issue by conal** [conal/concat#40]
Hm. With ghc 8.2.2, it looks like the "built-in rule" for `toCcc'` (doing most of the transformation work) never gets invoked. Investigating.

**Comment by cartazio** [comment]
Is there a way I can help?

**Comment by cartazio** [comment]
On the flip side it does make compiling the examples wayyyy faster :)

**Comment by conal** [comment]
Thanks to help from @kosmikus, I think I found the problem here.

In `ConCat.Plugin` [from pull #37](https://github.com/conal/concat/pull/37/files#diff-53dee52c802e3fcc4ebac2aeca1f3628R2279):

``` haskell
isMonoTy :: Type -> Bool
isMonoTy (TyConApp _ tys)      = all isMonoTy tys
isMonoTy (AppTy u v)           = isMonoTy u && isMonoTy v
#if MIN_VERSION_GLASGOW_HASKELL(8,2,0,0)

#else
isMonoTy (ForAllTy (Anon u) v) = isMonoTy u && isMonoTy v
#endif
isMonoTy (LitTy _)             = True
isMonoTy _                     = False
```

The omitted test handles function types, which are the only types given as an argument to `toCcc'`. The `isMonoTy` thus always yields `False`, and so the rewriting fails every time.

I made a very simple fix: use the `FunTy` pattern synonym regardless of GHC version. The new definition:

``` haskell
isMonoTy :: Type -> Bool
isMonoTy (TyConApp _ tys) = all isMonoTy tys
isMonoTy (AppTy u v)      = isMonoTy u && isMonoTy v
isMonoTy (FunTy u v)      = isMonoTy u && isMonoTy v
isMonoTy (LitTy _)        = True
isMonoTy _                = False
```

This change fixes the immediate issue with 8.2.2. I'll open another issue for the next breakage.

**Comment by cartazio** [comment]
sounds like the bug was my own fault :)
